clover-core/src/main/java/com/atlassian/clover/instr/java/java.g

```diff
@@ -3026,11 +3026,33 @@ CHAR_LITERAL
  * String literals. In double quotes we can have:
  *  - a character escaped by a backslash, such as '\t', '\u0000'
  *  - any other character except: CR, LF, double quote, non-escaped backslash
+ *
+ * In the text blocks, we can have any characters except for the three double
+ * quotes not led with a backslash.
  */
 STRING_LITERAL
+    : STRING_LITERAL_SINGLE_LINE
+    | STRING_LITERAL_TEXT_BLOCK
+    ;
+
+protected STRING_LITERAL_SINGLE_LINE
     : {nc();}   '"' ( ESC | ~( '"' | '\\' | '\n' | '\r') )* '"'
     ;
 
+protected STRING_LITERAL_TEXT_BLOCK
+    : {nc();} '"' '"' '"' ( '\r' | '\n' ) 
+        (   (  BACKSLASH   '"' '"'  '"' ) =>  BACKSLASH   '"' '"' '"'
+          | ( (BACKSLASH)? '"' '"' ~'"' ) => (BACKSLASH)? '"' '"'
+          | ( (BACKSLASH)? '"'     ~'"' ) => (BACKSLASH)? '"'
```

Comment:
Why do we need these semantic predicates and three variants of the backslash and quotes? Could you explain this a little bit, please?

Reply:
I have reviewed these lines in a little more detail and found the first line is unnecessary, so I have deleted the first line.
The second and third line is necessary to handle the code like this (embedded in `Java15TextBlock.java`):
```java
        System.out.println("""
                "" " " ""
                """);
```

Without the second and third lines, you will get an error like this:
```
ERROR: Instrumentation errorInstrumentation error : /mnt/c/home/satob/git/clover/clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java:10:19:expecting '"', found ' '
com.atlassian.clover.api.CloverException: /mnt/c/home/satob/git/clover/clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java:10:19:expecting '"', found ' '
```

Without semantic predicates, you will also get a similar error on the same line:
```
ERROR: Instrumentation errorInstrumentation error : /mnt/c/home/satob/git/clover/clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java:14:19:expecting '"', found ' '
com.atlassian.clover.api.CloverException: /mnt/c/home/satob/git/clover/clover-core/src/test/resources/javasyntax1.15/java15/Java15TextBlock.java:14:19:expecting '"', found ' '
```